wgpu-core/src/error.rs

```diff
@@ -0,0 +1,175 @@
+use core::fmt;
+use std::error::Error;
+
+use crate::{
+    gfx_select,
+    hub::{Global, IdentityManagerFactory},
+};
+
+pub struct ErrorFormatter<'a> {
+    writer: &'a mut dyn fmt::Write,
+    global: &'a Global<IdentityManagerFactory>,
+}
+
+impl<'a> ErrorFormatter<'a> {
+    pub fn error(&mut self, err: &dyn Error) {
+        writeln!(self.writer, "    {}", err).expect("Error formatting error");
+    }
+
+    pub fn note(&mut self, note: &dyn fmt::Display) {
```

Comment:
can this work with `FormatArgs`?

Reply:
Do you mean `fmt::Arguments`? And that the note/error/label stuff would return it instead of writing directly to the `Writer`? I don't really see the point, the users (PrettyError impls) would just pass it on directly to a `Writer` anyway?

Or do you mean instead of taking a `&dyn fmt::Display` argument, take a `&fmt::Arguments`? I suppose thats ok, though the caller would need to do some extra `format_args!`, that is not needed now?

Reply:
Ooh, do you mean to avoid the subsequent formatting, collapsing them to a single one, by passing fmt::Arguments until the note thing that writes it out?

Reply:
yeah, but you are right, it doesn't matter too much